### ISSUE TITLE
Allow actions at (x, y) to return objectID on success

### ISSUE THREAD

**Issue by Tushar-N** [allenai/ai2thor#483]
I'm using actions that accept x, y positions as inputs (e.g. `event = controller.step('OpenObject', x=0.2, y=0.5)`). Once it's executed, I'm trying to figure out what object was interacted with.

1. Is there any way to do it with the currently exposed methods?

2. Alternately, is there a way to just get the object that the raycast would hit? 
```
public bool ScreenToWorldTarget(float x, float y, ref SimObjPhysics target, bool requireWithinViewportRange)
```
does this and populates `target`, but this isn't an exposed method. Could this be made a callable easily?

Is there an easy way to do what I'm after? I thought I could just modify the interaction functions to return the target as a feedback parameter: e.g.`actionFinished(true) --> actionFinished(true, feedback)` like some methods that return info in `actionReturn`, but I'm finding it hard to parse recent code. PutObject has an `actionFinished(true)`, but OpenObject does not, so I got confused.

**Comment by thomason-jesse** [comment]
This appears to be implemented now.
[ https://ai2thor.allenai.org/ithor/documentation/actions/object-interaction/#pickup-object ]

It can also be done using the instance segmentation mask, as seen here generalized to an array of pixel values (performs IOU between provided pixel mask and instance segmentation to select the target object)
[ https://github.com/askforalfred/alfred/blob/master/env/thor_env.py#L490 ]

**Comment by thomason-jesse** [comment]
Did this feature ever pan out?

It would be possible to detect the altered object by comparing the objects' metadata before and after the action (for successfully executed actions), but that's computationally pretty tedious.

**Comment by mattdeitke** [comment]
I don't necessarily think returning the `objectId` will work in every case, since some actions (i.e., `TouchThenApplyForce`) already return information to the `actionReturn`.

What I can add is a query action to the environment for the object at `(x, y)`:
```python
controller.step('GetObject', x=x, y=y, forceAction=True/False)
```
where `forceAction` specifies if visibility/interactability distance to the object should be ignored.

If an object is there, the `objectId` is returned inside of `actionReturn`. If not, the action return will be either null or empty. Then, if an `objectId` is returned, you can guarantee that calling:
```python
controller.step('OpenObject', x=x, y=y, forceAction=True/False)
```
targets an object, or can equivalently pass in the returned objectId, which is marginally faster:
```python
objectId = controller.step('GetObject', x=x, y=y, forceAction=True/False).metadata['actionReturn']
if objectId:
    controller.step('OpenObject', objectId=objectId, forceAction=True/False)
```
Is this of interest?

**Comment by thomason-jesse** [comment]
If the proposed `GetObject` action would use the same raycast logic, that would do the trick for me!

**Comment by Tushar-N** [comment]
> What I may add is a query action to the environment for the object at (x, y):

The way I'm currently doing this is mapping from `event.instance_segmentation_frame` to `objectId` using `color_to_object_id`. Will `GetObject` be any faster thank this process? Since the raycast is happening for `OpenObject` anyway, I imagined getting this output for free, but now with two `controller.step(...)` calls, that may not be true?

Of course, it's still convenient to have an exposed `GetObject` method regardless.

**Comment by mattdeitke** [comment]
> but now with two controller.step(...) calls, that may not be true?

There are a few new big speedups (being announced in a couple of days) that make query actions (that don't change the state of the environment) significantly faster by not re-sending any of the metadata that won't change or re-sending all the image frames.

> The way I'm currently doing this is mapping from event.instance_segmentation_frame to objectId using color_to_object_id. Will GetObject be any faster than this process?

I expect that it'd be significantly faster, because of the new speedups and because rendering instance segmentation frames is quite slow.

**Comment by aishwaryap** [comment]
I would also like to have an exposed `GetObject` method. Using the segmentation is not a perfect solution - sometimes the action will still take effect on some object event of the segmentation method doesn't map the (x, y) to a valid object for the specified action.

**Comment by mattdeitke** [comment]
This functionality is now supported in v2.7.4. with the new `GetObjectInFrame` query action:

```python
forceAction = True
query = controller.step('GetObjectInFrame', x=x, y=y, forceAction=forceAction)
if query:
    controller.step(
        action='OpenObject',
        objectId=query.metadata['actionReturn'],
        forceAction=forceAction
    )
```

Documentation will be updated shortly. But, briefly, `forceAction=True` lets the targeted object be further away than the initialized visibility/insteractable distance of 1.5 meters.

Also note that `bool(ai2thor.server.Event)` (i.e., `bool(query)`) is now an alias to `query.metadata['lastActionSuccess']`.

---

If you're using Python 3.8+ you can also use the new walrus operator:
```python
forceAction = True
if (query :=  = controller.step('GetObjectInFrame', x=x, y=y, forceAction=forceAction):
    controller.step(
        action='OpenObject',
        objectId=query.metadata['actionReturn'],
        forceAction=forceAction
    )
```
which merges the query declaration inside of the conditional.

**Comment by thomason-jesse** [comment]
This is super helpful.

Is there any chance we could get a related function like `GetXYZinFrame` that just returns the simulator RaycastHit endpoint in (x, y, z) sim space given an (x, y) frame coordinate? In tandem with the recent `GetSpawnCoordinatesAboveReceptacle` and `PlaceObjectAtPoint` functions, such functionality would also let me write a workaround for https://github.com/allenai/ai2thor/issues/540
